### PR TITLE
Refactor BinaryBuild architecture and add option to build from file

### PR DESCRIPTION
**Pull request by istraka**
@simkam @pkremens 

Please make sure your PR meets the following requirements:
- [x] Pull Request contains a description of the changes
- [x] Pull Request does not include fixes for multiple issues/topics
- [x] Code is formatted, imports ordered, code compiles and tests are passing
- [x] Code is self-descriptive and/or documented

### CODE REVIEW

File: core/src/main/java/cz/xtf/core/bm/BinaryBuild.java [hunk 1]
```diff
@@ -56,7 +28,7 @@
 
 @Slf4j
 public abstract class BinaryBuild implements ManagedBuild {
-	private static final String CONTENT_HASH_LABEL_KEY = "xtf.bm/content-hash";
+	static final String CONTENT_HASH_LABEL_KEY = "xtf.bm/content-hash";
```
Comment (simkam) [inline core/src/main/java/cz/xtf/core/bm/BinaryBuild.java:31]: what is the reason for this change?

File: core/src/main/java/cz/xtf/core/bm/BinaryBuild.java [hunk 2]
```diff
@@ -66,10 +38,10 @@
 	private final String builderImage;
 	private final Map<String, String> envProperties;
 
-	private final ImageStream is;
-	private final BuildConfig bc;
+	final ImageStream is;
```
Comment (simkam) [inline core/src/main/java/cz/xtf/core/bm/BinaryBuild.java:41]: Shouldn't it be `protected` so any subclass can see these fields?

File: core/src/main/java/cz/xtf/core/bm/BinaryBuild.java [hunk 3]
```diff
@@ -81,26 +53,6 @@ public BinaryBuild(String builderImage, Path path, Map<String, String> envProper
 		this.bc = this.createBcDefinition();
 	}
 
-	@Override
-	public void build(OpenShift openShift) {
```
Comment (simkam) [inline core/src/main/java/cz/xtf/core/bm/BinaryBuild.java:85]: this breaks API, no? It should be at least replaced with abstract method

File: core/src/main/java/cz/xtf/core/bm/BinaryBuild.java [hunk 4]
```diff
@@ -239,48 +191,5 @@ private BuildConfig createBcDefinition() {
...
-					}
-				}
-				o.closeArchiveEntry();
-			}
-			o.finish();
-		} catch (ArchiveException | IOException e) {
-			throw new RuntimeException(e);
-		}
-	}
+	abstract String getContentHash();
```
Comment (simkam) [inline core/src/main/java/cz/xtf/core/bm/BinaryBuild.java:194]: `protected` ?

File: core/src/main/java/cz/xtf/core/bm/BinaryBuild.java (hunk 1 above)
Comment (istraka) [reply core/src/main/java/cz/xtf/core/bm/BinaryBuild.java:31]: Hash code is computed in subclasses now. I will change it to `protected`.

File: core/src/main/java/cz/xtf/core/bm/BinaryBuild.java (hunk 3 above)
Comment (istraka) [reply core/src/main/java/cz/xtf/core/bm/BinaryBuild.java:85]: Yeah good catch. If an object typed as `BinaryBuld` instead of `ManagedBuild` then yes.

**Comment by istraka** [comment]
@simkam I have fixed issues. Moreover I made all abstract method protected so they are is easily overridden.

File: core/src/main/java/cz/xtf/core/bm/BinaryBuild.java [hunk 5]
```diff
@@ -238,49 +192,4 @@ private BuildConfig createBcDefinition() {
...
-				contentHash = Hex.encodeHexString(dos.getMessageDigest().digest()).substring(0, 63);
-			} catch (NoSuchAlgorithmException e) {
-				throw new RuntimeException(e);
-			}
-		}
-
-		return contentHash;
-	}
-
-	private void writeProjectTar(OutputStream os) {
```
Comment (mnovak1) [inline core/src/main/java/cz/xtf/core/bm/BinaryBuild.java:260]: `BinaryBuild` is abstract class inherited in `BinaryDockerBuild`, `BinarySourceBuild` and `MutableBinarySourceBuild`. I don't like that if `writeProjectTar` and  `getContentHash` are removed here in `BinaryBuild` then it duplicated in inhereted classes. I'm not sure what is the reason for change.

Comment (istraka) [reply core/src/main/java/cz/xtf/core/bm/BinaryBuild.java:260]: It has been moved into `BinaryBuildFromSources`.

Comment (mnovak1) [reply core/src/main/java/cz/xtf/core/bm/BinaryBuild.java:260]: Ok, got it.
